File: java/src/hackerrank/BeautifulTriplets.java
package hackerrank;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.assertEquals;

// https://www.hackerrank.com/challenges/beautiful-triplets/problem
public class BeautifulTriplets {

    static int beautifulTriplets_Solution2(int d, int[] arr) {
        List<Integer> list = Arrays.stream(arr).boxed().collect(Collectors.toList());
        return list.stream().reduce(0, (acc, n) -> (list.contains(n+d) && list.contains(n+2*d)) ? acc+1 : acc);
    }

    static int beautifulTriplets_Solution1(int d, int[] arr) {
        int result = 0;
        List<Integer> list = Arrays.stream(arr).boxed().collect(Collectors.toList());
        for (int n: list) {
            if (list.contains(n+d) && list.contains(n+2*d))
                result++;
        }
        return result;
    }


    @Test
    @DisplayName("Test sample input 0")
    void beautifulTriplets0() {
        int[] arr = {1, 2, 4, 5, 7, 8, 10};
        int d = 3, expected = 3;
        int result2 = beautifulTriplets_Solution2(d, arr);
        int result1 = beautifulTriplets_Solution1(d, arr);
        assertEquals(expected, result2);
    }

}

File: kotlin/src/hackerrank/beautifulTriplets.kt
package hackerrank

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test


// https://www.hackerrank.com/challenges/beautiful-triplets/problem

fun beautifulTriplets_Solution2(d: Int, arr: Array<Int>): Int {
    return arr.fold(0) { acc, n -> if (arr.contains(n + d) && arr.contains(n + 2 * d)) acc+1 else acc }
}

fun beautifulTriplets_Solution1(d: Int, arr: Array<Int>): Int {
    var result = 0
    for (n in arr) {
        if (arr.contains(n + d) && arr.contains(n + 2 * d))
            result++
    }
    return result
}


internal class BeautifulTripletsTest {

    @Test
    @DisplayName("Test sample 0")
    fun beautifulTriplets0() {
        val arr: Array<Int> = arrayOf(1, 2, 4, 5, 7, 8, 10)
        val d = 3; val expected = 3
        val result2 = beautifulTriplets_Solution2(d, arr)
        val result1 = beautifulTriplets_Solution1(d, arr)
        assertEquals(expected, result2)
        assertEquals(expected, result1)
    }

}